src/main/java/com/rcv/Tabulator.java

```diff
@@ -234,58 +234,70 @@ private void logSummaryInfo() {
   // logic only considers continuing candidates, so it won't assign any votes to past winners -- but
   // in reality they continue to hold their winning margins for the rest of the rounds, so we need
   // to fill in those values here.
-  // TODO: instead of recomputing these winning tallies each round, we could just read the values
-  // from the previous round. Once someone wins and we redistribute their extra votes, their tally
-  // should be frozen for the rest of the rounds.
+  // We need to do the computation once for each winner to account for transfers. In subsequent
+  // rounds, we can just copy the number from the previous round, since it won't change.
   private void updatePastWinnerTallies() {
-    // pastWinners contains winners from rounds that preceded the current round
-    Set<String> pastWinners = new HashSet<>();
+    Map<String, BigDecimal> roundTally = roundTallies.get(currentRound);
+    Map<String, BigDecimal> previousRoundTally = roundTallies.get(currentRound - 1);
+    List<String> winnersToProcess = new LinkedList<>();
+    Set<String> winnersRequiringComputation = new HashSet<>();
     for (String winner : winnerToRound.keySet()) {
       // skip someone who won in the current round, because we already have that tally filled in
-      if (winnerToRound.get(winner) < currentRound) {
-        pastWinners.add(winner);
+      int winningRound = winnerToRound.get(winner);
+      if (winningRound == currentRound) {
+        continue;
+      }
+      winnersToProcess.add(winner);
+      if (winningRound == currentRound - 1) {
```

Comment:
I'm a little confused here: why do all this checking for the previous round, etc., instead of just keeping a _global_ hash table of winners, and skipping calcs for them if they're in that list.  I realize this is similar to what you had before (except that was done locally within the function); I don't quite understand what this here buys us that a global table wouldn't, and it seems that the latter would be much simpler.

Reply:
Perhaps I'm misunderstanding your question... but `winnerToRound` is an instance variable (which I think is what you meant by global). I'm just rearranging that data here for convenience. If I'm missing your point, could you elaborate?

Reply:
Ok, got it. I still don't feel like I've got a good enough grasp on everything here to explicitly approve it (I should probably stop trying to review code late at night), but no other objections for me.